_published/0013-authentication.md

```diff
@@ -0,0 +1,177 @@
+validate the token. See
+[validating an OIDC Access Token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens).
+
+### Session Management
+
+The wharf-api is currently stateless and does not use sessions. To keep this
+design tokens will be sent in every request to the server.
+
+The option of introducing sessions as a way to store tokens may be reconsidered
+at a later point. There seems to be a potential performance gain.
+
+### Local development vs production environments
+
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+For use in offline developer environments it should be possible to switch on a
+developer mode switch that allows bypassing validation of access
+tokens such that a valid login is not required.
+
+
+### SSL Config Required as a part of OIDC
+
+For OIDC ssl is required. SSL certs will be inserted into the traefik config
+and generated with openSSL as part of the repo setup.
+
+### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. See
+[validating an access token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens)
+
+## Compatibility
+
+Expect no compatibility between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a wasted opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different IDPs. A sample situation where this could be useful is
+when an external third party should be granted view or admin rights to the
+deployment pipeline.
+
+When handling tokens from RBAC enabled IDPs it's possible to verify
+custom permission levels using custom role claims.
+[See this link](https://docs.microsoft.com/en-us/azure/role-based-access-control/custom-roles)
```

Comment:
The page has a warning message:

> Adding a management group to AssignableScopes is currently in preview. This preview version is provided without a service level agreement, and it's not recommended for production workloads. /.../
>
> https://docs.microsoft.com/en-us/azure/role-based-access-control/custom-roles

Is this something we need to worry about? Have you tested this feature, or is it hypothetical?

(You're not required to have tested this feature, as it's for a future plan and not of major importence right now. I'm merely curious)

Reply:
I have tested this in parts. Its hidden behind the azure premium paywall. Its executable through auth0/okta iaas.